feat(develop/span-first): Add implementation guidelines page (#15717)
getsentry/sentry-docs: develop-docs/sdk/telemetry/spans/implementation.mdx

---
title: Implementation Guidelines
sidebar_order: 10
---

<Alert level="warning">
  🚧 This document is work in progress.
  The steps and suggestions in this document primarily serve as a means to document what SDKs so far have been doing when implementing Span-First.
  This page also serves as a place to document (temporary) decisions, trade-offs, considerations, etc.
</Alert>

<Alert>
  This document uses key words such as "MUST", "SHOULD", and "MAY" as defined in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt) to indicate requirement levels.
</Alert>

This document provides guidelines for implementing Span-First in SDKs. This is purposefully NOT a full specification. For exact specifications, refer to the other pages under [Spans](..).

## How To Approach Span-First in SDKs

If you're implementing Span-First (as a PoC) in your SDK, take an iterative approach in which you implement the functionality incrementally. Here's a rough suggestion for iterations.

1. Add the Span v2 Envelope (type), serialization logic and any utilities necessary to support sending a new envelope. See [Span Protocol](../span-protocol) for more details.

> 2. Add the top-level `traceLifeCycle` (or `trace_life_cycle`) SDK init option which controls if traces should be sent as transactions or as spans (v2).

Author: ```suggestion
2. Add the top-level `traceLifecycle` (or `trace_lifecycle`) SDK init option which controls if traces should be sent as transactions or as spans (v2).
```

>    - The allowed values for this option MUST be `'static'` and `'streamed'`.

Author: ```suggestion
   - The allowed values for this option MUST be `'static'` and `'strea'`.
```

   - By default, the SDK MUST send traces as transactions (`'static'`). Span-First MUST be an opt-in feature.

>    - Continue with adding Span-First logic which MUST only be applied if `traceLifeCycle` is set to `'streamed'`.

Author: ```suggestion
   - Continue with adding Span-First logic which MUST only be applied if `traceLifeCycle` is set to `'stream'`.
```

## Span Buffer

This section is intentionally short because all buffering specification is being added to the [Telemetry Buffer](../../telemetry-buffer) page.

Some rough pointers:
- Given that SDKs SHOULD materialize and freeze the DSC as late as possible, the span buffer SHOULD enqueue span instances and at _flush time_ serialize them to JSON.
  Before serialization, the span buffer SHOULD materialize and freeze the DSC on the segment span if not already done so.
  This ensures that the `trace` envelope header has the most up to date data from the DSC (e.g. relevant for `transaction` names in the DSC).
- SDKs SHOULD follow one of the backend, mobile or browser telemetry buffer specifications. 
- It is expected and fine to implement the proper, weighted buffering logic as a final step in the Span-First project. 
  Intermediate buffers MAY be simpler, for example disregard the priority logic and just buffer until a certain span length, size or time interval is reached.
## Release

The initial PoC implementation of Span-First **SHOULD** be released in a **minor version** of the SDK.

> - This feature is entirely opt-in via `traceLifeCycle = 'streamed'` and therefore does **not** introduce breaking changes to existing users.

Author: ```suggestion
- This feature is entirely opt-in via `traceLifeCycle = 'stream'` and therefore does **not** introduce breaking changes to existing users.
```
